packages/block-editor/src/components/list-view/block-select-button.js

```diff
@@ -61,30 +67,51 @@ function ListViewBlockSelectButton(
 		context: 'list-view',
 	} );
 	const { isLocked } = useBlockLock( clientId );
-	const { canToggleBlockVisibility, isBlockHidden, hasPatternName } =
+	const { canToggleBlockVisibility, hasPatternName, blockVisibility } =
 		useSelect(
 			( select ) => {
 				const { getBlockName, getBlockAttributes } =
 					select( blockEditorStore );
-				const { isBlockHidden: _isBlockHidden } = unlock(
-					select( blockEditorStore )
-				);
 				const blockAttributes = getBlockAttributes( clientId );
 				return {
 					canToggleBlockVisibility: hasBlockSupport(
 						getBlockName( clientId ),
 						'visibility',
 						true
 					),
-					isBlockHidden: _isBlockHidden( clientId ),
 					hasPatternName: !! blockAttributes?.metadata?.patternName,
+					blockVisibility: blockAttributes?.metadata?.blockVisibility,
 				};
 			},
 			[ clientId ]
 		);
 	const shouldShowLockIcon = isLocked;
+	// Show visibility icon when block has any visibility settings (regardless of current viewport)
 	const shouldShowBlockVisibilityIcon =
-		canToggleBlockVisibility && isBlockHidden;
+		canToggleBlockVisibility && hasAnyVisibilitySettings( blockVisibility );
+
+	// Compute visibility tooltip text
+	const getVisibilityTooltip = () => {
+		if ( blockVisibility === false ) {
+			return __( 'Hidden' );
+		}
+		if ( typeof blockVisibility === 'object' ) {
+			const hiddenDevices = [
+				blockVisibility.desktop === false && __( 'Desktop' ),
+				blockVisibility.tablet === false && __( 'Tablet' ),
+				blockVisibility.mobile === false && __( 'Mobile' ),
+			].filter( Boolean );
+			if ( hiddenDevices.length ) {
+				return sprintf(
+					/* translators: %s: comma-separated list of device types */
+					__( 'Hidden on %s' ),
+					hiddenDevices.join( ', ' )
```

Comment:
Sometimes we join with " and ", sometimes with a comma. Should we be consistent?